File: Programmers_Rotate_Matrix_Borders.js
function solution(rows, columns, queries) {
    var answer = [];
    let board = [];
    
    for(let i = 0; i < rows + 1; i++) {
        let len = board.length;
        let base = i - 1;
        
        board[len] = []
        for(let j = 0; j < columns + 1; j++) {
            board[len][j] = base * columns + j;
        }
    }
    
    for(let i = 0; i < queries.length; i++) {
        const [y1, x1, y2, x2] = queries[i];
        let stack = [];
        
        for(let i = x1; i < x2; ++i) stack.push(board[y1][i]);
        for(let i = y1; i < y2; ++i) stack.push(board[i][x2]);
        for(let i = x2; i > x1; --i) stack.push(board[y2][i]);
        for(let i = y2; i > y1; --i) stack.push(board[i][x1]);
        
        answer.push(Math.min(...stack))
        stack.unshift(stack.pop());
        
        for(let i = x1; i < x2; ++i) board[y1][i] = stack.shift();
        for(let i = y1; i < y2; ++i) board[i][x2] = stack.shift();
        for(let i = x2; i > x1; --i) board[y2][i] = stack.shift();
        for(let i = y2; i > y1; --i) board[i][x1] = stack.shift();
    }
    
    return answer;
}

File: Programmers_Rotate_Matrix_Borders.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    int board[rows+1][columns+1];
    
    for(int i = 0; i < rows; i++) {
        int base = columns * i;
        
        for(int j = 1; j < columns + 1; j++) {
            board[i][j] = base + j;
        }
    }
    
    for(int i = 0; i < queries.size(); i++) {
        vector<int> stack;
        
        int x1 = queries[i][0] - 1;
        int y1 = queries[i][1];
        int x2 = queries[i][2] - 1;
        int y2 = queries[i][3];
        
        for(int k = y1; k < y2; k++) stack.push_back(board[x1][k]);
        for(int k = x1; k < x2; k++) stack.push_back(board[k][y2]);
        for(int k = y2; k > y1; k--) stack.push_back(board[x2][k]);
        for(int k = x2; k > x1; k--) stack.push_back(board[k][y1]);
        
        int tail = stack[stack.size()-1];
        int min_tail = tail;
        stack.pop_back();
        
        for(int k = x1 + 1; k < x2; k++) {
            int tmp = stack[stack.size()-1];
            board[k][y1] = tmp;
            min_tail = min_tail > tmp ? tmp : min_tail;
            stack.pop_back();
        }
        
        for(int k = y1; k < y2; k++) {
            int tmp = stack[stack.size()-1];
            board[x2][k] = tmp;
            min_tail = min_tail > tmp ? tmp : min_tail;
            stack.pop_back();
        }
        
        for(int k = x2; k > x1; k--) {
            int tmp = stack[stack.size()-1];
            board[k][y2] = tmp;
            min_tail = min_tail > tmp ? tmp : min_tail;
            stack.pop_back();
        }
        
        for(int k = y2; k > y1; k--) {
            int tmp = stack[stack.size()-1];
            board[x1][k] = tmp;
            min_tail = min_tail > tmp ? tmp : min_tail;
            stack.pop_back();
        }
        
        board[x1][y1] = tail;
        answer.push_back(min_tail);
    }
    
    return answer;
}